src/textual/_resolve.py

```diff
@@ -78,13 +79,107 @@ def resolve(
+    widget_styles: Iterable[RenderStyles],
+    size: Size,
+    viewport_size: Size,
+    remaining_space: Fraction,
+    resolve_dimension: Literal["width", "height"] = "width",
+) -> Fraction:
+    """Calculate the fraction
+
+    Args:
+        widget_styles: Styles for widgets with fraction units.
+        size: Container size.
+        viewport_size: Viewport size.
+        remaining_space: Remaining space for fr units.
+        resolve_dimension: Which dimension to resolve.
+
+    Returns:
+        The value of 1fr.
+    """
+    if not remaining_space or not widget_styles:
+        return Fraction(1)
+
+    def resolve_scalar(
+        scalar: Scalar | None, fraction_unit: Fraction = Fraction(1)
+    ) -> Fraction | None:
+        """Resolve a scaler if it is not None.
+
+        Args:
+            scalar: Optional scalar to resolve.
+            fraction_unit: Size of 1fr.
+
+        Returns:
+            Fraction if resolved, otherwise None.
+        """
+        return (
+            None
+            if scalar is None
+            else scalar.resolve(size, viewport_size, fraction_unit)
+        )
+
+    resolve: list[tuple[Scalar, Fraction | None, Fraction | None]] = []
+
+    if resolve_dimension == "width":
+        resolve = [
+            (
+                cast(Scalar, styles.width),
+                resolve_scalar(styles.min_width),
+                resolve_scalar(styles.max_width),
+            )
+            for styles in widget_styles
+        ]
+    else:
+        resolve = [
+            (
+                cast(Scalar, styles.height),
+                resolve_scalar(styles.min_height),
+                resolve_scalar(styles.max_height),
+            )
+            for styles in widget_styles
+        ]
+
+    resolved: list[Fraction | None] = [None] * len(resolve)
+    remaining_fraction = Fraction(sum(scalar.value for scalar, _, _ in resolve))
+
+    while True:
+        remaining_space_changed = False
+        for index, (scalar, min_value, max_value) in enumerate(resolve):
+            value = resolved[index]
+            if value is None:
+                resolved_scalar = scalar.resolve(
+                    size, viewport_size, Fraction(remaining_space, remaining_fraction)
+                )
+                if min_value is not None and resolved_scalar < min_value:
+                    remaining_space -= min_value
+                    remaining_fraction -= Fraction(scalar.value)
+                    resolved[index] = min_value
+                    remaining_space_changed = True
+                elif max_value is not None and resolved_scalar > max_value:
+                    remaining_space -= max_value
+                    remaining_fraction -= Fraction(scalar.value)
+                    resolved[index] = max_value
+                    remaining_space_changed = True
+                else:
+                    resolved[index] = resolved_scalar
+        if not remaining_space_changed:
+            break
```

Comment:
In the first pass of this loop, `resolved` will _not_ contain a single `None` and the second iteration will never do anything.

In a related note, the order in which the widgets are passed into this function can affect the final result.
If this is intentional, I would suggest documenting it.

Consider a vertical container with height 20 and these three children with `fr` height:

 - A: `height: 1fr`, `min-height: 4`
 - B: `height: 2fr`, `min-height: 14`
 - C: `height: 1fr`

Depending on whether the widgets are presented in this order or if you flip the first two, widget `A` may or may not trigger the code that checks that its minimum height is greater than the space reserved for `fr` units, which in turns affects the values of `remaining_space` and `remaining_fraction`.

I also think this can be done in a _single_ pass (maybe two, one for the max and another for the min) if we sort the widgets correctly but I would have to spend more time thinking about this to figure out if my intuition is right, or not.